torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -274,9 +281,19 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
   auto requestId = nextId();
   auto future = std::make_shared<FutureMessage>();
   if (message.isRequest()) {
+    // millisecond level precision of when request started.
+    auto futureStartTime =
+        std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::steady_clock::now().time_since_epoch());
     {
       std::lock_guard<std::mutex> lock{futureMutex_};
-      futures_[requestId] = future;
+      futures_[requestId] = std::make_pair(future, futureStartTime);
+      // insert future into timeouts map to keep track of its timeout
+      if (futureTimeouts_.find(futureStartTime) != futureTimeouts_.end()) {
+        futureTimeouts_[futureStartTime].emplace_back(requestId);
+      } else {
+        futureTimeouts_[futureStartTime] = std::vector<int64_t>{requestId};
+      }
```

Comment:
Why the two different paths?

The `operator[]` function will always return a reference. If the value doesn't yet exist it is default constructed. Instead of first finding if an entry with the start time already exists, you can simply do:

```c++
futureTimeouts_[futureStartTime].push_back(requestId);
```

Also no need to use `emplace_back` here because we're dealing with a trivial type.